### PR TITLE
Added Support for SoftPlus operator in tflite.

### PR DESCRIPTION
**Pull request by amitsrivastava78**
This is part of issue #27822.

### CODE REVIEW

**Comment by jdduke** [comment]
Do you have a specific model which uses this operator?

We're trying to raise the bar for adding new operators to TFLite, preferring a smaller set of core operators where possible, as it makes it more difficult to maintain parity with our delegate/accelerator pipeline.

**Comment by amitsrivastava78** [comment]
@jdduke , thanks for the comments, this implementation is inline with the issue #27822. But i understand your point. So  should we use graph transform to convert softplus operator or leave it for the custom ops ?
As i feel graph transform might again not be so efficient as it will create again two nodes(when value is within the Range). Let me know your take on this.

Regards
Amit

**Comment by jdduke** [comment]
I'm going to defer to @miaout17 for further review. I agree that the graph transform won't be quite as important. Can you tell if the SoftPlus operator in that graph is used sparingly? Or all throughout?

**Comment by amitsrivastava78** [comment]
@jdduke , thanks for the response, I think the model uses softplus sparingly, but softplus seems to be picking up momentum.
@miaout17 , could you please review and provide the feedback.

Regards
Amit

**Comment by gbaned** [comment]
@amitsrivastava78 could you please resolve the conflicts?

**Comment by amitsrivastava78** [comment]
@gbaned , thanks for pointing this out, i have resolved the conflicts.

Regards
Amit

File: tensorflow/lite/kernels/register.cc [hunk 1]
```diff
@@ -356,6 +357,9 @@ BuiltinOpResolver::BuiltinOpResolver() {
   AddBuiltin(BuiltinOperator_NOT_EQUAL, Register_NOT_EQUAL(),
              /* min_version */ 1,
              /* max_version */ 2);
+  AddBuiltin(BuiltinOperator_SOFTPLUS, Register_SOFTPLUS(),
+             /* min_version */ 1,
```
Comment (miaout17) [inline tensorflow/lite/kernels/register.cc]: These aren't necessary. Softplus is a new op so it has only one version

File: tensorflow/lite/kernels/softplus.cc [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+#include "tensorflow/lite/kernels/kernel_util.h"
+#include "tensorflow/lite/kernels/op_macros.h"
+
+namespace tflite {
+namespace ops {
+namespace builtin {
+namespace softplus {
+
+struct OpData {
+  float threshhold;
```
Comment (miaout17) [inline tensorflow/lite/kernels/softplus.cc]: typo? (fix all "threshhold" in the file)

File: tensorflow/lite/kernels/softplus.cc [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+  TfLiteTensor* output;
+};
+
+TfLiteStatus Prepare(TfLiteContext* context, TfLiteNode* node) {
+  TF_LITE_ENSURE_EQ(context, NumInputs(node), 1);
+  TF_LITE_ENSURE_EQ(context, NumOutputs(node), 1);
+
+  OpData* data = reinterpret_cast<OpData*>(node->user_data);
+  // threshhold for avoiding the underflow and overflow problems.
+  data->threshhold = std::log(std::numeric_limits<float>::epsilon()) + 2.0f;
```
Comment (miaout17) [inline tensorflow/lite/kernels/softplus.cc]: This can simply be a constant under softplus namespace

**Review by miaout17 — COMMENTED** [review]
Thanks for sending the change!

**Comment by miaout17** [comment]
While the code itself looks good, the following questions are unanswered yet:

* What models use SoftPlus op? 
* Should we do graph transformation or define a new op? (tradeoff between efficiency versus fewer builtin ops). 

@jdduke are you comfortable adding this as a builtin op? If yes, I can follow up the code review process

File: tensorflow/lite/kernels/register.cc (hunk 1 above)
Comment (amitsrivastava78) [reply tensorflow/lite/kernels/register.cc]: @miaout17 , thanks for the review i have updated the comments as per your suggestion.

Regards
Amit

File: tensorflow/lite/kernels/softplus.cc (hunk 2 above)
Comment (amitsrivastava78) [reply tensorflow/lite/kernels/softplus.cc]: @miaout17 , thanks for the review i have updated the comments as per your suggestion.

Regards
Amit

File: tensorflow/lite/kernels/softplus.cc (hunk 3 above)
Comment (amitsrivastava78) [reply tensorflow/lite/kernels/softplus.cc]: @miaout17 , thanks for the review i have updated the comments as per your suggestion.

Regards
Amit

**Comment by amitsrivastava78** [comment]
@miaout17 thanks for the review i have updated all the comments as per your suggestion, kindly check.


Regards
Amit

**Comment by jdduke** [comment]
Hi @amitsrivastava78, we're working on some guidelines for new operators, could you give us a few days to get back to you?

As noted previously, it would be good to know if other models are using this operator, and whether a different activation would suffice.

**Comment by amitsrivastava78** [comment]
@jdduke , thanks for your response, sure i will wait for the conclusion from your side, also i will check which all models are using this operator and update you.

Regards
Amit

**Comment by gbaned** [comment]
@amitsrivastava78 could you please resolve the conflicts? Thanks!

**Comment by amitsrivastava78** [comment]
@gbaned , thanks for pointing this out, i have resolved the conflicts.

Regards
Amit

**Comment by gbaned** [comment]
@amitsrivastava78 could you please resolve the conflicts? Thanks!

**Comment by amitsrivastava78** [comment]
@gbaned , i have rebased the code and resolved the merge conflicts.

@jdduke can you please have a look at the PR and let me know your feedback

Regards
Amit

**Comment by jdduke** [comment]
Until we find another model which requires this op, I'd rather we rely on using the select TF ops, or we add this as a custom op that users can optionally link into their app.

**Comment by jdduke** [comment]
See also https://github.com/tensorflow/tensorflow/commit/03195f13456354deea8b81c9e583621b1337b952#diff-2b45693b554369bde8c98e9a76b80036

**Comment by gbaned** [comment]
@amitsrivastava78 Could you please address the jdduke's comments. Thanks!

**Comment by rthadur** [comment]
I'm going to go ahead and close this PR, because it seems to have stalled. If you're still interested in pursing this (and responding to my comments), please feel free to reopen!

**Comment by divyajaincs** [comment]
Does Tflite support SoftPlus operator now, if not how can i include that as a custom one. @jdduke

**Pull request closed without merging**